### PR TITLE
perf: Cache staging buffer in structured output to fix memory regression (#49013)

### LINKED ISSUE #49013: [Perf] ~2x decode throughput regression for structured outputs since #45424: apply_grammar_bitmask staging rewrite (bisected to commit, file, and hunk)
## Summary

End-to-end decode throughput for guided-JSON (xgrammar) workloads drops ~2x
between v0.23.0 and v0.24.0 (also v0.25.1). Bisected to commit
**7df3d7dada840c68b85b26b79de7f59f676d58e3** "[Core] Ensure memory is pinned
prior to async h2d copy" (#45424), then within the commit to a single file:
**vllm/v1/structured_output/utils.py — `apply_grammar_bitmask`**.

Evidence chain (identical harness, deterministic to ±1s per 2,000-request
drain; A100 80GB, Gemma-4-26B-A4B-IT bf16, max_num_seqs=64, ngram spec decode
k=5, CUDA graphs, EVERY request uses output_format json_schema → xgrammar):

| build | drain time | rate |
|---|---|---|
| 8dd1b702f (parent commit) | 513 s | 233 req/min |
| 7df3d7dad (culprit) | 985 s | 121 req/min |
| 7df3d7dad + ALL 29 v1/* files reverted to parent | 534 s | 224 req/min |
| 7df3d7dad + ONLY structured_output/utils.py reverted | 534 s | 224 req/min |
| 7df3d7dad + worker files (gpu_model_runner/gpu_input_batch) reverted | 1006 s | 119 req/min |
| 7df3d7dad + PIN_MEMORY forced False (torch_utils) | 985 s | 121 req/min |
| v0.24.0 release | 7581 s vs 3810 s on 15k corpus | 2.0x slower |

Note the PIN_MEMORY=False row: forcing the new global pin flag off does NOT
recover throughput, so the cost is in the staging rewrite itself, not merely
the pinned allocation flag.
Hunk-level split: restoring ONLY the staging block below (keeping the
commit's other change in the same function — index_tensor via
async_tensor_h2d) fully recovers baseline: **514 s / 233 req/min**. The
staging rewrite is the entire regression.

## The suspect code

#45424 changed the per-step bitmask staging in `apply_grammar_bitmask` from:

```python
sorted_bitmask = np.full(
    shape=(logits.shape[0], grammar_bitmask.shape[1]), fill_value=-1,
    dtype=grammar_bitmask.dtype,
)
...  # reorder loop fills rows
grammar_bitmask = torch.from_numpy(sorted_bitmask).to(logits.device, non_blocking=True)
```

to:

```python
sorted_bitmask_tensor = torch.full(
    (logits.shape[0], grammar_bitmask.shape[1]), -1,
    dtype=torch.from_numpy(grammar_bitmask[:0]).dtype,
    pin_memory=PIN_MEMORY,
)
sorted_bitmask = sorted_bitmask_tensor.numpy()
...  # same reorder loop, now writing into the (pinned) tensor's numpy view
grammar_bitmask = sorted_bitmask_tensor.to(logits.device, non_blocking=True)
```

For a 262k-vocab model at max_num_seqs=64 with k=5 ngram speculation this is a
~(384 x 8192) int32 = ~12.6 MB freshly allocated (normally pinned) host tensor
per engine step, plus the reorder loop writing through the numpy view of that
pinned buffer, freed again at the end of the step. This path runs every step
for every batch containing structured-output requests.

## Setup detail

- A100 80GB PCIe, torch 2.11.0 (cu13), python 3.12, offline LLM engine
  (in-process), xgrammar backend, chunked prefill 8192, TRITON_ATTN
  (forced by Gemma4 hetero head dims), async scheduling disabled (ngram).
- Symptom profile: engine steps uniformly ~2x slower at identical batch
  occupancy, acceptance rates, KV/prefix-cache hit, GPU clocks; CUDA graphs
  on/off preserves the gap; regression also present WITHOUT spec decode
  (0.25.1 no-ngram run degraded >2x as well). All requests guided-JSON.
- Release-level numbers (14,985-request production-derived corpus):
  v0.21.0 3810 s makespan / 531 accepted tok/s → v0.24.0 7581 s / 266,
  v0.25.1 7590 s / 265.

## Suggested fix

Restore the numpy staging (or keep a cached, reused pinned staging buffer
sized to (max_logits_rows, bitmask_width) instead of a fresh per-step
allocation).

Possibly related: #47047 (Gemma-4-26B-A4B on SM120/NVFP4, 0.21→0.23 slowdown
at low concurrency, unresolved — different window but same model family and
also structured-output-heavy workloads).

Full bisect logs (9 commit-level probes via per-commit wheels + 8 file/hunk-level
image probes) available on request.

### PR DESCRIPTION
**Pull request by Tejas-Raj01**
## Purpose
Fixes #49013.

This PR resolves a ~2x throughput regression in structured output decoding. The bottleneck was identified as high-frequency allocations of ~12MB pinned host tensors inside `vllm/v1/structured_output/utils.py` during the `apply_grammar_bitmask` step. 

To fix this, the PR implements a cached `_STAGING_BUFFER`. This buffer is allocated once, reused across steps to eliminate allocation overhead, and includes logic to dynamically resize only when the required tensor shape increases.

## Test Plan
Added a new unit test (`tests/v1/test_structured_output_perf.py`) to verify the caching logic. The test validates:
1. Initial staging buffer allocation.
2. Buffer reuse without reallocation when the input shape remains the same.
3. Successful dynamic resizing when the request batch size or bitmask width increases.

**Test Command:**
```bash
python -m pytest tests/v1/test_structured_output_perf.py -v

### CODE REVIEW

**Comment by Tejas-Raj01** [comment]
Hi @mgoin, @russellb, @aarnphm, and @benchislett,

I would also appreciate it if anyone could take a look at the changes when you have a moment. Thank you.

**Comment by Tejas-Raj01** [comment]
@mgoin @russellb @aarnphm @benchislett @njhill

ready for review.

**Review by chaunceyjiang — COMMENTED** [review]
Could you provide a benchmark for this?

**Comment by Tejas-Raj01** [comment]
Hi @chaunceyjiang,

I have benchmarked the performance impact of the grammar_staging_cache. As expected, caching the buffer eliminates the reallocation overhead on subsequent steps. Here are the precise metrics from the structured output performance test:

Initial Allocation Time: 106.3590 ms

Buffer Reuse Time: 0.0639 ms

Dynamic Resizing Time: 0.1053 ms

The staging cache successfully avoids the ~106ms overhead during generation steps. Let me know if you need anything else.

File: vllm/v1/structured_output/utils.py [hunk 1]
```diff
@@ -109,36 +117,51 @@ def apply_grammar_bitmask(
...
-        if (logit_idx := struct_out_req_batch_indices.get(req_id)) is not None:
-            for i in range(1 + num_spec_tokens):
-                bitmask_index = logit_idx + i
-                sorted_bitmask[bitmask_index] = grammar_bitmask[cumulative_index + i]
-                out_indices.append(bitmask_index)
-        cumulative_index += 1 + num_spec_tokens
+        req_batch_idx = struct_out_req_batch_indices[req_id]
+
+        for i in range(num_spec_tokens + 1):
+            sorted_bitmask[req_batch_idx + i] = grammar_bitmask[cumulative_index]
```
[the remark this answers is not in the document]
Comment (Tejas-Raj01) [reply vllm/v1/structured_output/utils.py]: Good catch! I have applied the suggested fix. 

The `cumulative_index` is now correctly manually tracked and advances by `num_spec_tokens + 1` per request. Additionally, the inner loop now properly indexes the distinct rows for speculative positions using `grammar_bitmask[cumulative_index + i]`. 

The changes have been pushed and this should resolve the failing CI checks related to speculative decoding. Thank you for pointing this out.

**Comment by njhill** [comment]
@Tejas-Raj01 please try https://github.com/vllm-project/vllm/pull/49919 to see if it fixes the regression for you.

**Comment by Tejas-Raj01** [comment]
Hi @njhill,

I ran local benchmarks to verify the fix and compare the performance of my staging buffer implementation in this PR against the threading changes in #49919. 

Here are the results using `Qwen/Qwen2.5-0.5B-Instruct` generating 100 prompts (10,000 total generated tokens) in my local TP=1 environment:

| Metric | This PR (#49168) | PR #49919 |
| :--- | :--- | :--- |
| **Total Time** | 3.66 seconds | 2.55 seconds |
| **Throughput** | 2734.72 tokens/s | 3915.89 tokens/s |

The results confirm that #49919 effectively resolves the regression and provides a significantly higher throughput in my setup. 

Let me know how you would like to proceed. Should I close this PR, or would it be valuable to eventually combine my memory staging cache on top of your threading fix once it is merged?

**Comment by njhill** [comment]
Thanks very much for running that @Tejas-Raj01, that's great to hear! I think we can close this one since we are aiming to keep the code as simple as possible and are in the process of deprecating the V1 model runner anyhow.
